### PR TITLE
ENH: init to dimension of x matrix rather than rank

### LINKED ISSUE #2597: quantreg: exception with singular exog
I'm trying to run quantile regression with a number of variables.  It appears `quantreg.fit()` will not allow more than 8 variables/columns to be included in the formula.  

```
import pandas as pd 
import statsmodels.formula.api as smf
from statsmodels.regression.quantile_regression import QuantReg

d = pd.read_csv('<path-to-data>')

#7 vars (lag and sunday through friday), seems to work fine
qr = smf.quantreg('load ~ lag + su + m + t + w + th + f',  d')
qm = qr.fit(q=0.5)

#8 vars (lag and sunday through saturday), throws error
qr = smf.quantreg('load ~ lag + su + m + t + w + th + f + s',  d')
qm = qr.fit(q=0.5)
```

Here's the trace:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-17-f2074b49225f> in <module>()
---> 32 qm = qr.fit(q=.5)

.../statsmodels/regression/quantile_regression.py in fit(self, q, vcov, kernel, bandwidth, max_iter, p_tol, **kwargs)
    177             resid = np.abs(resid)
    178             xstar = exog / resid[:, np.newaxis]
--> 179             diff = np.max(np.abs(beta - beta0))
    180             history['params'].append(beta)
    181             history['mse'].append(np.mean(resid*resid))

ValueError: operands could not be broadcast together with shapes (9,) (8,) 
```

### PR DESCRIPTION
**Pull request by jlpolit**
- [x] closes #2597
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message). 

Updating quantile regression to initialize using the dimension of the X matrix rather than the rank. This might break things, so testing before finalizing the PR. Will add a test or three if current test suite passes with the change.

**Notes**:

* It is essential that you add a test when making code changes. Tests are not 
  needed for doc changes.
* When adding a new function, test values should usually be verified in another package (e.g., R/SAS/Stata).
* When fixing a bug, you must add a test that would produce the bug in master and
  then show that it is fixed with the new code.
* New code additions must be well formatted. Changes should pass flake8. If on Linux or OSX, you can
  verify you changes are well formatted by running 
  ```
  git diff upstream/master -u -- "*.py" | flake8 --diff --isolated
  ```
  assuming `flake8` is installed. This command is also available on Windows 
  using the Windows System for Linux once `flake8` is installed in the 
  local Linux environment. While passing this test is not required, it is good practice and it help 
  improve code quality in `statsmodels`.
* Docstring additions must render correctly, including escapes and LaTeX.

</details>

### CODE REVIEW

**Comment by josef-pkt** [comment]
closes #2597

File: statsmodels/regression/tests/test_quantile_regression.py [hunk 1]
```diff
@@ -1,9 +1,9 @@
 import scipy.stats
 import numpy as np
-import statsmodels.api as sm
 from numpy.testing import assert_allclose, assert_equal, assert_almost_equal
-from patsy import dmatrices  # pylint: disable=E0611
+import statsmodels.api as sm
 from statsmodels.regression.quantile_regression import QuantReg
+from patsy import dmatrices  # pylint: disable=E0611
```
Comment (bashtage) [inline statsmodels/regression/tests/test_quantile_regression.py]: Ideal order is

Std library

External packages

statsmodels


and within each block imports before from.

import a
import b
from c import d
from e import f

File: statsmodels/regression/tests/test_quantile_regression.py [hunk 2]
```diff
@@ -281,3 +281,15 @@ def test_alpha_summary():
...
+
+
+def test_collinear_matrix():
+    X = np.array([[1, 0, .5], [1, 0, .8],
+                  [1, 0, 1.5], [1, 0, .25]], dtype=np.float64)
+    y = np.array([0, 1, 2, 3], dtype=np.float64)
+
+    try:
+        est_collinear = QuantReg(y, X).fit(0.5)
+    except Exception as e:
```
Comment (josef-pkt) [inline statsmodels/regression/tests/test_quantile_regression.py]: what exception does this raise in the singular case?

using rank  before the fix was an obvious bug, 
However, it is not clear how the singular exog case works or should work

Comment (jlpolit) [reply statsmodels/regression/tests/test_quantile_regression.py]: yeah that's not 100% clear to me yet (this PR was in part aimed to check whether unit tests passed with the change, and start the conversation...still a work in progress)

when I run with this change with a singular design matrix, the only obvious thing that comes up is that line 199 (` h = bandwidth(nobs, q)`) comes up as nan with the default bandwidth, which raises a RuntimeWarning but no Exception. 

It seems like at minimum an appropriate warning or error should be raised so the user knows that they passed a singular design matrix to the model, but I'm also not certain how much IRLS is impacted by a singular matrix

Comment (josef-pkt) [reply statsmodels/regression/tests/test_quantile_regression.py]: > ( h = bandwidth(nobs, q)) comes up as nan

this sounds strange, this bandwidth does not depend on the data, just on nobs

**Comment by josef-pkt** [comment]
behavior at singular exog

IRLS uses pinv, so we get similar behavior to OLS and IRLS in GLM (*), with a pinv regularized solution
`beta = np.dot(pinv(xtx), xty)`
(QuantReg uses pinv on product matrix x'x instead of on exog directly)

So, we could verify that pinv-regularized `fit` works correctly, instead of raising an exception wherever it breaks.

File: statsmodels/regression/tests/test_quantile_regression.py (hunk 2 above)
Comment (jlpolit) [reply statsmodels/regression/tests/test_quantile_regression.py]: Ah! I misspoke - it's not bandwidth itself that comes up nan, but the call to `norm.ppf(q+h)` on line 200.
but changing the bandwidth does fix the problem....and I think it's because for some reason that default bandwidth is resulting in an h val that's greater than q, which leads us to take the pdf of numbers that are >1 and <0

**Comment by jlpolit** [comment]
> behavior at singular exog
> 
> IRLS uses pinv, so we get similar behavior to OLS and IRLS in GLM (*), with a pinv regularized solution
> `beta = np.dot(pinv(xtx), xty)`
> (QuantReg uses pinv on product matrix x'x instead of on exog directly)
> 
> So, we could verify that pinv-regularized `fit` works correctly, instead of raising an exception wherever it breaks.

based on the test provided, it at least appears to work without throwing an exception....but it might be preferable to have a nontrivial test to confirm correctness (particularly in light of the runtime warning mentioned in convo above)

**Comment by josef-pkt** [comment]
You could create an random sample with a repeated column (2 columns in X are identical) and a sample size, for example, nobs=100.
I usually just use this kind of random data to check how an estimator or statistic works with singular, perfectly collinear data.

As a cross check, dropping a perfectly collinear column should result in the same prediction, fittedvalues as including it with pinv regularization (IIRC).

File: statsmodels/regression/tests/test_quantile_regression.py (hunk 2 above)
Comment (josef-pkt) [reply statsmodels/regression/tests/test_quantile_regression.py]: Good, I guess this is then an issue with the small sample size that you use. There might be a minimum sample size that the bandwidth creates a valid value.
You could open a separate issue for that if I guess correctly

Comment (jlpolit) [reply statsmodels/regression/tests/test_quantile_regression.py]: ...yep, based on some quick-and-dirty testing it looks like this can have issues with relatively low sample size. I may open an issue in lieu of doing more rigorous testing to find that boundary.

**Comment by jlpolit** [comment]
...there seems to be a coverage test failing due to a change in `statsmodels/graphics/functional.py`, which I don't think is touched by this PR? did I miss something about what's going on with that test?

**Comment by josef-pkt** [comment]
I don't know what causes "weird" test coverage results. In my experience, we get sometimes unrelated changes in unit test coverage for things that haven't been changed. I guess it's a problem in how they compute changes in the code, or how test results are compared.
You can ignore that.

If singular exog doesn't raise an exception, then you can remove the `try except` in the unit tests.

**Comment by josef-pkt** [comment]
LGTM, Thanks for the fix and sorry for the delay.

12 commits is too much for this
Can you do a rebase and squash the commits?

**Comment by jlpolit** [comment]
> LGTM, Thanks for the fix and sorry for the delay.
> 
> 12 commits is too much for this
> Can you do a rebase and squash the commits?

No worries - thanks!

**Comment by josef-pkt** [comment]
rebased version with additional unit test merged in #7694

@jlpolit Thank you for the PR. I was slow in getting back to this. It will now be in 0.13
